### PR TITLE
Use firmware path from sysfs

### PR DESCRIPTION
**Pull request by minlexx**
Some linux distributions adjust path fro kernel to load firmware from using `/sys/module/firmware_class/parameters/path`. Kernel supports it, teach tqftpserv to respect it too.

Kernel supports only single firmware path specified in firmware_class path. In postmarketOS we created a service script [msm-firmware-loader](https://gitlab.com/postmarketOS/msm-firmware-loader/) that runs in early boot stage and symlinks all firmware it can find on various device partitions into a single directory, which is then written into  `/sys/module/firmware_class/parameters/path`.

With this patch tqftpserv is able, for example, to successfully serve `wlanmdsp.mbn` file when modem remoteproc is started.
```
$ sudo ./tqftpserv
[TQFTP] WRQ: /readwrite/server_check.txt (octet)
[TQFTP] RRQ: /readwrite/ota_firewall/ruleset (octet)
tqftpserv: failed to open ota_firewall/ruleset: No such file or directory
[TQFTP] unable to open /readwrite/ota_firewall/ruleset (2), reject
[TQFTP] RRQ: /readwrite/ota_firewall/ruleset (octet)
tqftpserv: failed to open ota_firewall/ruleset: No such file or directory
[TQFTP] unable to open /readwrite/ota_firewall/ruleset (2), reject
[TQFTP] RRQ: /readonly/firmware/image/wlanmdsp.mbn (octet)
[TQFTP] Remote returned an error: End of Transfer
[TQFTP] RRQ: /readonly/firmware/image/wlanmdsp.mbn (octet)
[TQFTP] Remote returned an error: End of Transfer
[TQFTP] RRQ: /readonly/firmware/image/wlanmdsp.mbn (octet)
```

Can be used to solve https://github.com/linux-msm/tqftpserv/issues/7

### CODE REVIEW

File: translate.c [hunk 1]
```diff
@@ -53,6 +53,20 @@
...
 
+static void read_fw_path_from_sysfs(char *outbuffer, size_t bufsize)
+{
+	size_t pathsize;
+	FILE *f = fopen("/sys/module/firmware_class/parameters/path", "rt");
+	if (!f)
+		return;
+	pathsize = fread(outbuffer, sizeof(char), bufsize, f);
+	fclose(f);
+	if (pathsize <= 0)
```
Comment (konradybcio) [inline translate.c:64]: not sure if it can be negative

File: translate.c [hunk 2]
```diff
@@ -111,6 +128,23 @@ static int translate_readonly(const char *file)
 		}
 		firmware_value[n] = '\0';
 
+		/* first try path from sysfs */
+		if ((strlen(fw_sysfs_path) > 0) &&
+		    (strlen(fw_sysfs_path) + 1 +strlen(firmware_value) + 1 + strlen(file) + 1 < sizeof(path))) {
```
Comment (konradybcio) [inline translate.c:133]: space after `+ 1 +`

File: translate.c [hunk 3]
```diff
@@ -111,6 +128,23 @@ static int translate_readonly(const char *file)
...
+			strcpy(path, fw_sysfs_path);
+			strcat(path, "/");
+			strcat(path, dirname(firmware_value));
+			strcat(path, "/");
+			strcat(path, file);
```
Comment (lumag) [inline translate.c:138]: This looks like a perfect usecase for `asprintf`

File: translate.c (hunk 1 above)
Comment (minlexx) [reply translate.c:64]: indeed return value type of fread is size_t, so `== 0` it is

File: translate.c (hunk 2 above)
Comment (minlexx) [reply translate.c:133]: :astonished: oops

File: translate.c (hunk 3 above)
Comment (minlexx) [reply translate.c:138]: asprintf/vanprintf is non-standard GNU/BSD extension and existing code around already uses strcpy/strcat everywhere. I can try though to use it here if you insist

Comment (lumag) [reply translate.c:139]: No, I don't.

**Review by lumag — APPROVED** [review]

**Pull request merged**